### PR TITLE
Support for unstructured text corpus datasets for CPT

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses: #845, #809 

Continued pre-training involves an identical data processing pipeline as standard pre-training, where the model simply predicts the next token and completes the text. It does not require any templating or prompt formatting. Our existing dataset classes are specifically designed for instruct and chat tuning, but don't support free-form text corpuses.

Here, we add a dataset class `TextDataset` that simply calls `load_dataset` and tokenizes the text directly without any further processing. It uses `encode` which simply adds BOS/EOS tokens if needed. This should be compatible with llama2, llama3, and other model's formatting requirements.

#### Changelog
What are the changes made in this PR?
- Add `TextDataset` and appropriate tests
- Add an example dataset builder, `cnn_dailymail_articles_dataset` and appropriate tests

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Ran the full finetune distributed recipe with the cnn dailymail dataset. 
```
2024-04-24:23:12:06,234 INFO     [_utils.py:34] Running FullFinetuneRecipeDistributed with resolved config:

batch_size: 2
checkpointer:
  _component_: torchtune.utils.FullModelMetaCheckpointer
  checkpoint_dir: /tmp/Meta-Llama-3-8B/original/
  checkpoint_files:
  - consolidated.00.pth
  model_type: LLAMA3
  output_dir: /tmp/Meta-Llama-3-8B/
  recipe_checkpoint: null
dataset:
  _component_: torchtune.datasets.cnn_dailymail_articles_dataset
device: cuda
dtype: bf16
enable_activation_checkpointing: true
epochs: 3
gradient_accumulation_steps: 1
log_every_n_steps: null
loss:
  _component_: torch.nn.CrossEntropyLoss
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.utils.metric_logging.WandBLogger
  log_dir: /tmp/alpaca-llama3-finetune
  name: cnn_dailymail
  project: torchtune
model:
  _component_: torchtune.models.llama3.llama3_8b
optimizer:
  _component_: torch.optim.AdamW
  foreach: false
  lr: 2.0e-05
output_dir: /tmp/alpaca-llama3-finetune
resume_from_checkpoint: false
seed: null
shuffle: true
tokenizer:
  _component_: torchtune.models.llama3.llama3_tokenizer
  path: /tmp/Meta-Llama-3-8B/original/tokenizer.model
```

image

### CODE REVIEW

File: torchtune/datasets/_text_completion.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+from datasets import load_dataset
+from torch.utils.data import Dataset
+from torchtune.data import truncate
+from torchtune.modules.tokenizers import Tokenizer
+
+
+class TextCompletionDataset(Dataset):
+    """
+    Freeform dataset for any unstructured text corpus. Quickly load any dataset
```
Comment (rohan-varma) [inline torchtune/datasets/_text_completion.py:17]: what does "correctly" mean here :D

File: torchtune/datasets/_text_completion.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        tokens = self._tokenizer.encode(text=prompt, add_bos=True, add_eos=True)
+
+        # Truncate if needed, but don't coerce EOS id
+        if self.max_seq_len is not None:
+            tokens = truncate(tokens, self.max_seq_len - 1)
+
+        # No need to offset labels by 1 - happens in the recipe
+        labels = tokens.copy()
+
+        assert len(tokens) == len(labels)
```
Comment (rohan-varma) [inline torchtune/datasets/_text_completion.py:62]: this seems pretty trivial as we do a copy right above. Worth removing?

File: tests/torchtune/datasets/test_text_completion_dataset.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+from tests.test_utils import DummyTokenizer
+
+from torchtune.datasets import TextCompletionDataset
+
+
+class TestTextCompletionDataset:
+    expected_tokenized_prompts = [
+        [0, 4, 2, 2, 7, 5, -1],
+        [0, 4, 2, 7, 7, 5, -1],
```
Comment (rohan-varma) [inline tests/torchtune/datasets/test_text_completion_dataset.py:17]: Worth using BOS and EOS token constants directly from a tokenizer here?

**Review by rohan-varma — COMMENTED** [review]
This is cool, do we in general plan to support CPT in torchtune and plan for an RFC around this possibly?

File: tests/torchtune/datasets/test_text_completion_dataset.py (hunk 3 above)
Comment (RdoubleA) [reply tests/torchtune/datasets/test_text_completion_dataset.py:17]: Yeah I suppose using the DummyTokenizer's bos/eos directly would be better

**Comment by RdoubleA** [comment]
> do we in general plan to support CPT in torchtune and plan for an RFC around this possibly?
I don't know that CPT will be sufficiently different from our finetune recipes to warrant its own, besides different data handling. @rohan-varma would you know :) or maybe @pbontrager ?

**Comment by pbontrager** [comment]
> I don't know that CPT will be sufficiently different from our finetune recipes to warrant its own, besides different data handling. @rohan-varma would you know :) or maybe @pbontrager ?

I don't think so initially. I think this dataset is good and we could add a recipe in the future if there are additional custom things that people want for advanced use cases.

File: recipes/configs/llama3/8B_full.yaml [hunk 4]
```diff
@@ -72,8 +72,10 @@ dtype: bf16
 
 # Logging
 metric_logger:
-  _component_: torchtune.utils.metric_logging.DiskLogger
+  _component_: torchtune.utils.metric_logging.WandBLogger
```
Comment (joecummings) [inline recipes/configs/llama3/8B_full.yaml:75]: bad rafi

Comment (RdoubleA) [reply recipes/configs/llama3/8B_full.yaml:75]: ohno

File: recipes/configs/llama3/8B_full.yaml [hunk 5]
```diff
@@ -25,8 +25,8 @@ tokenizer:
 
 # Dataset
 dataset:
-  _component_: torchtune.datasets.alpaca_dataset
-  train_on_input: True
+  _component_: torchtune.datasets.wikitext_dataset
```
Comment (joecummings) [inline recipes/configs/llama3/8B_full.yaml:28]: also don't change the default here

File: torchtune/datasets/_cnn_dailymail.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        TextCompletionDataset: the configured TextCompletionDataset
+    """
+
+    return TextCompletionDataset(
+        tokenizer=tokenizer,
+        source=source,
+        column="article",
+        max_seq_len=max_seq_len,
+        split="train",
+        name="3.0.0",
```
Comment (joecummings) [inline torchtune/datasets/_cnn_dailymail.py:43]: What is "name"?

**Comment by joecummings** [comment]
Any update here? I want to train on https://huggingface.co/datasets/allenai/dolma soon.

File: recipes/configs/llama3/8B_full.yaml (hunk 5 above)
Comment (RdoubleA) [reply recipes/configs/llama3/8B_full.yaml:28]: ohno

File: torchtune/datasets/_cnn_dailymail.py (hunk 6 above)
Comment (RdoubleA) [reply torchtune/datasets/_cnn_dailymail.py:43]: let me add a comment, but it's basically specifying the subset of data

File: torchtune/datasets/_text_completion.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        if self.max_seq_len is not None:
+            tokens = truncate(tokens, self.max_seq_len - 1)
+
+        # No need to offset labels by 1 - happens in the recipe
+        labels = tokens.copy()
+
+        return tokens, labels
+
+
+def text_completion_dataset(
```
Comment (joecummings) [inline torchtune/datasets/_text_completion.py:65]: Sorry one last nit: Can you add an examples section like this? https://github.com/pytorch/torchtune/blob/29ae975fc6d2f8e85ce33634116e0bda0472253c/torchtune/datasets/_chat.py#L137

Comment (joecummings) [reply torchtune/datasets/_text_completion.py:65]: It'll make the docs and usage much clearer

**Review by joecummings — APPROVED** [review]
🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥
